pkg/logcli/query/query.go

```diff
@@ -157,9 +195,195 @@ func (q *Query) DoQuery(c client.Client, out output.LogOutput, statistics bool)
 				// fudge the timestamp forward in time to make sure to get the last entry from this batch in the next query
 				end = lastEntry[0].Timestamp.Add(1 * time.Nanosecond)
 			}
+		}
+	}
+
+	if partFile != nil {
+		if err := partFile.Complete(); err != nil {
+			log.Fatalln(err)
+		}
+	}
+}
+
+func (q *Query) outputFilename() string {
+	return fmt.Sprintf("%s-%d-%d.part", q.PartPrefix, q.Start.Unix(), q.End.Unix())
+}
+
+// createPartFile returns a PartFile if the PartFilePrefix is set.
+// The bool value shows if the part file already exists, and this range should be skipped.
+func (q *Query) createPartFile() (*PartFile, bool) {
+	if q.PartPrefix == "" {
+		return nil, false
+	}
+
+	partFile := NewPartFile(q.outputFilename())
+
+	if !q.OverwriteCompleted {
+		// If we already have the completed file, no need to download it again.
+		// The user can delete the files if they want to download parts again.
+		exists, err := partFile.Exists()
+		if err != nil {
+			log.Fatalf("Query failed: %s\n", err)
+		}
+		if exists {
+			log.Printf("Skip range: %s - %s: already downloaded\n", q.Start, q.End)
+			return nil, true
+		}
+	}
+
+	if err := partFile.CreateTemp(); err != nil {
+		log.Fatalf("Query failed: %s\n", err)
+	}
+
+	return partFile, false
+}
+
+// rounds up duration d by the multiple m, and then divides by m.
+func durationCeilDiv(d, m time.Duration) int64 {
+	return int64((d + m - 1) / m)
+}
+
+// Returns the next job's start and end times.
+func (q *Query) nextJob(start, end time.Time) (time.Time, time.Time) {
+	if q.Forward {
+		start = end
+		return start, minTime(start.Add(q.ParallelDuration), q.End)
+	}
+
+	end = start
+	return maxTime(end.Add(-q.ParallelDuration), q.Start), end
+}
+
+type parallelJob struct {
+	q    *Query
+	done chan struct{}
+}
+
+func newParallelJob(q *Query) *parallelJob {
+	return &parallelJob{
+		q:    q,
+		done: make(chan struct{}),
+	}
+}
 
+func (j *parallelJob) run(c client.Client, out output.LogOutput, statistics bool) {
+	j.q.DoQuery(c, out, statistics)
```

Comment:
I'm wondering if you can miss some log entries if the time range has more logs than the configured limit.

Reply:
Ok. I've finally got the answer. Say the limit is 1000.
- if the first job fetches 1000 lines it's done and all remaining jobs can be canceled
- if the job gets less than 1000 lines the next need to continue

Reply:
`DoQuery` is always [limited](https://github.com/grafana/loki/blob/main/pkg/logcli/query/query.go#L52) by [--limit](https://github.com/grafana/loki/blob/main/cmd/logcli/main.go#L365). Say we have the limit 100 and a query for an hour and thus four 15 minutes jobs.

If the first job reaches 100 entries the processing should stop. Right now this implementation will return more entries a the other jobs will return up to 100 entries as well. They might also have gaps as the following example should illustrate.

Say we time ranges 12:00-12:15, 12:15-12:30, 12:30-12:45 and 12:45-13:00. End is exclusive. If 12:00-12:15 has actually 150 entries `DoQuery` will only process the first 100. The job for 12:15-12:30 will not includes these.

Reply:
I suggest we abide to the limit and follow up with `--limit=-1` or something similar to disable it.

Reply:
Right. The reason this worked for me is because I was setting the limit to something really high. I did this long ago and forgot why I did that.

Reply:
I've pushed a commit, it's using the same solution used [here](https://github.com/grafana/loki/blob/main/cmd/logcli/main.go#L172)

Reply:
I think a proper way to illustrate an infinite limit would be to have a struct with an `Infinite/NegativeInfinite bool` parameters with `GreaterThan` and `LessThan` functions taking the infinite values into account. Kind of similar to how SQL handles this. But it means refactoring all the usages of `Limit`, and testing it. That could be a bit of work. Will someone have more than 2^32 log lines? I mean, maybe it's possible. No one thought ipv4 would fill up when they made it.

Reply:
> I've pushed a commit, it's using the same solution used here

That is a great solution. I agree that the limit won't make much sense with parallel execution. Please add the documentation I've proposed.

Reply:
Sure thing. I will update the docs tomorrow.